BraintreeUIKit/Models/BTUIKCardType.m

```diff
@@ -63,40 +63,17 @@ + (instancetype)cardTypeForNumber:(NSString *)number {
         return nil;
     }
     for (BTUIKCardType *cardType in [[self class] allCards]) {
-        for (NSString *prefix in cardType.validNumberPrefixes) {
-            if (number.length >= prefix.length) {
-                NSUInteger compareLength = MIN(prefix.length, number.length);
-                NSString *sizedNumber = [number substringToIndex:compareLength];
-                if ([sizedNumber isEqualToString:prefix]) {
-                    return cardType;
-                }
-            }
-        }
-    }
-    return nil;
-}
+        for (NSString *pattern in cardType.validNumberPrefixes) {
+            NSError  *error = nil;
+            NSRegularExpression* regex = [NSRegularExpression regularExpressionWithPattern:pattern options:0 error:&error];
+            NSTextCheckingResult *match = [regex firstMatchInString:number options:0 range: NSMakeRange(0, [number length])];
 
-// Since each card type has a list of acceptable card prefixes, we
-// can determine which card types may match a given number
-+ (NSArray *)possibleCardTypesForNumber:(NSString *)number {
-    number = [BTUIKUtil stripNonDigits:number];
-    if (number.length == 0) {
-        return [[self class] allCards];
-    }
-    NSMutableSet *possibleCardTypes = [NSMutableSet set];
-    for (BTUIKCardType *cardType in [[self class] allCards]) {
-        for (NSString *prefix in cardType.validNumberPrefixes) {
-            NSUInteger compareLength = MIN(prefix.length, number.length);
-
-            NSString *sizedPrefix = [prefix substringToIndex:compareLength];
-            NSString *sizedNumber = [number substringToIndex:compareLength];
-            if ([sizedNumber isEqualToString:sizedPrefix]) {
-                [possibleCardTypes addObject:cardType];
-                break;
+            if (match != nil) {
```

Comment:
I believe we talked about this in another PR. I'm going to move forward with this check. Also,`location` is not a property of the `match`.